huff0/decompress_amd64.go

```diff
@@ -8,21 +8,52 @@ package huff0
 import (
 	"errors"
 	"fmt"
+
+	"github.com/klauspost/compress/internal/cpuinfo"
 )
 
 // decompress4x_main_loop_x86 is an x86 assembler implementation
 // of Decompress4X when tablelog > 8.
 // go:noescape
-func decompress4x_main_loop_x86(pbr0, pbr1, pbr2, pbr3 *bitReaderShifted,
+func decompress4x_main_loop_amd64(pbr0, pbr1, pbr2, pbr3 *bitReaderShifted,
+	peekBits uint8, buf *byte, tbl *dEntrySingle) uint8
+
+// decompress4x_main_loop_x86 is an x86 assembler implementation
+// of Decompress4X when tablelog > 8.
+// go:noescape
+func decompress4x_main_loop_bmi2(pbr0, pbr1, pbr2, pbr3 *bitReaderShifted,
+	peekBits uint8, buf *byte, tbl *dEntrySingle) uint8
+
+type decompress4x_main_loop_fn func(pbr0, pbr1, pbr2, pbr3 *bitReaderShifted,
+	peekBits uint8, buf *byte, tbl *dEntrySingle) uint8
+
+var decompress4x_main_loop decompress4x_main_loop_fn
+var decompress4x_8b_main_loop decompress4x_main_loop_fn
+
+// decompress4x_8b_loop_x86 is an x86 assembler implementation
+// of Decompress4X when tablelog <= 8 which decodes 4 entries
+// per loop.
+// go:noescape
+func decompress4x_8b_main_loop_amd64(pbr0, pbr1, pbr2, pbr3 *bitReaderShifted,
 	peekBits uint8, buf *byte, tbl *dEntrySingle) uint8
 
 // decompress4x_8b_loop_x86 is an x86 assembler implementation
 // of Decompress4X when tablelog <= 8 which decodes 4 entries
 // per loop.
 // go:noescape
-func decompress4x_8b_loop_x86(pbr0, pbr1, pbr2, pbr3 *bitReaderShifted,
+func decompress4x_8b_main_loop_bmi2(pbr0, pbr1, pbr2, pbr3 *bitReaderShifted,
 	peekBits uint8, buf *byte, tbl *dEntrySingle) uint8
 
+func init() {
+	if cpuinfo.HasBMI2() {
+		decompress4x_main_loop = decompress4x_main_loop_bmi2
+		decompress4x_8b_main_loop = decompress4x_8b_main_loop_bmi2
+	} else {
+		decompress4x_main_loop = decompress4x_main_loop_amd64
+		decompress4x_8b_main_loop = decompress4x_8b_main_loop_amd64
```

Comment:
This could be the cause of your slowdown.

These will be seen as escaping, since `decompress4x_main_loop` and `decompress4x_8b_main_loop ` no longer promises that arguments don't escape. That is why I made [this change](https://github.com/klauspost/compress/pull/537/files#diff-908cbe713b5231ca740e9771bfafd3cf23d764df659078bd7419633ab1e55219R64-R68) as well.

Try branching and calling directly, instead of having these `var ... = func()...`

Reply:
Thank you, will check it on Thursday. Last week I couldn't carry on, as got some cold and was recovering for a few days, ugh.

Reply:
>  Last week I couldn't carry on, as got some cold and was recovering for a few days, ugh.

Same here. GF got Covid and I started having symptoms Friday. Let's hope we both get well soon :)

Reply:
Keep my fingers crossed for you!